### ISSUE TITLE
Save nothing when using modelcheckpoint callbacks

### ISSUE THREAD

**Issue by 1zeryu** [Lightning-AI/pytorch-lightning#19339]
### Bug description

Using modelcheckpoint callbacks to save model, but save nothing.
trainer: config
  trainer:
    _target_: pytorch_lightning.Trainer
    _convert_: all
    max_epochs: 30000
    check_val_every_n_epoch:
    log_every_n_steps: 10
    val_check_interval : 3000
    accelerator: gpu
    devices:
    - ${device.id}
    limit_val_batches: 1

cwd = os.path.join(self.cfg.output_dir, cfg.system.name, time.strftime("%Y-%m-%d-%H-%M-%S", time.localtime()))
        checkpoint = ModelCheckpoint(monitor='loss', mode='min', save_top_k=1, save_last=True, every_n_train_steps=500, save_on_train_epoch_end=True,
                                     filename='{epoch}-{loss:.4f}')
        checkpoint2 = ModelCheckpoint(monitor='ae_acc', mode='max', save_top_k=1, save_last=False, every_n_train_steps=500, save_on_train_epoch_end=True,
                                      filename='{epoch}-{ae_acc:.4f}')
        trainer: Trainer = hydra.utils.instantiate(self.cfg.system.train.trainer, logger=TensorBoardLogger(cwd, '.', '.'),
                                                enable_model_summary=False,
                                               callbacks=[checkpoint, checkpoint2])
        trainer.fit(self.system, datamodule=datamodule)  

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
Here is my code:
--- BaseSystem
class BaseSystem(pl.LightningModule):
    def __init__(self, cfg, **kwargs):
        super(BaseSystem, self).__init__()
        self.save_hyperparameters()
        # when save  hyperparameters, the self.task will be ignored
        task_cfg =  cfg.task
        self.automatic_optimization = False
        self.config = cfg.system
        self.automatic_optimization = False
        self.train_cfg = self.config.train
        self.model_cfg = self.config.model
        self.model = self.build_model()
        self.loss_func = self.build_loss_func()
        self.data_transform = self.build_data_transform()
        self.build_task(task_cfg)
        pass

    def build_task(self, task_cfg, **kwargs):
        self.task = tasks[task_cfg.name](task_cfg, **kwargs)

    def get_task(self):
        return self.task

    def build_data_transform(self):
        if 'data_transform' in self.model_cfg and self.model_cfg.data_transform is not None:
            return hydra.utils.instantiate(self.config.data_transform)
        else:
            return None

    def build_trainer(self):
        trainer = hydra.utils.instantiate(self.train_cfg.trainer)
        pdb.set_trace()
        return trainer

    def task_func(self, input):
        return self.task.test_g_model(input)

    def training_step(self, batch, batch_idx, **kwargs):
        loss = self.forward(batch, **kwargs)
        self.optimizer.zero_grad()
        self.manual_backward(loss)
        self.optimizer.step()

        if hasattr(self, 'lr_scheduler'):
            self.lr_scheduler.step()

    def build_model(self, **kwargs):
        model = hydra.utils.instantiate(self.model_cfg.arch)
        return model

    def build_loss_func(self):
        if 'loss_func' in self.train_cfg:
            loss_func = hydra.utils.instantiate(self.train_cfg.loss_func)
            return loss_func
        else:
            warnings.warn("No loss function is specified, using default loss function")


    def configure_optimizers(self, **kwargs):
        params = self.model.parameters()
        self.optimizer = hydra.utils.instantiate(self.train_cfg.optimizer, params)

        if 'lr_scheduler' in self.train_cfg and self.train_cfg.lr_scheduler is not None:
            self.lr_scheduler = hydra.utils.instantiate(self.train_cfg.lr_scheduler)

        return self.optimizer

    def validation_step(self, batch, batch_idx, **kwargs):
        # TODO using task layer
        pass

    @abc.abstractmethod
    def forward(self, x, **kwargs):
        raise
--- Running System
class EncoderSystem(BaseSystem):
    def __init__(self, config, **kwargs):
        super(EncoderSystem, self).__init__(config, **kwargs)

    def forward(self, batch, **kwargs):
        output = self.model(batch)
        loss = self.loss_func(batch, output, **kwargs)
        # self.log('epoch', self.current_epoch)
        self.log('loss', loss.cpu().detach().mean().item(), on_epoch=True, prog_bar=True, logger=True)
        return loss

    def validation_step(self, batch, batch_idx, **kwargs):
        # todo
        good_param = batch[:10]
        input_accs = []
        for i, param in enumerate(good_param):
            acc, test_loss, output_list = self.task_func(param)
            input_accs.append(acc)
        print("input model accuracy:{}".format(input_accs))

        """
        AE reconstruction parameters
        """
        print('---------------------------------')
        print('Test the AE model')
        ae_rec_accs = []
        ae_params = self.decode(self.encode((good_param)))
        for i, param in enumerate(ae_params):
            acc, test_loss, output_list = self.task_func(param)
            ae_rec_accs.append(acc)

        best_ae = max(ae_rec_accs)
        print(f'AE reconstruction models accuracy:{ae_rec_accs}')
        print(f'AE reconstruction models best accuracy:{best_ae}')
        print('---------------------------------')
        self.log('ae_acc', best_ae)

    def encode(self, x, **kwargs):
        return self.model.encode(x)

    def decode(self, x, **kwargs):
        return self.model.decode(x)
--- Running
cfg = self.cfg
        self.system_cls = systems[cfg.system.name]
        self.system = self.system_cls(cfg, **kwargs)
        datamodule = self.system.get_task().get_param_data()
        # running
        self.output_dir = cfg.output_dir

        cwd = os.path.join(self.cfg.output_dir, cfg.system.name, time.strftime("%Y-%m-%d-%H-%M-%S", time.localtime()))
        checkpoint = ModelCheckpoint(monitor='loss', mode='min', save_top_k=1, save_last=True, every_n_train_steps=500, save_on_train_epoch_end=True,
                                     filename='{epoch}-{loss:.4f}')
        checkpoint2 = ModelCheckpoint(monitor='ae_acc', mode='max', save_top_k=1, save_last=False, every_n_train_steps=500, save_on_train_epoch_end=True,
                                      filename='{epoch}-{ae_acc:.4f}')
        trainer: Trainer = hydra.utils.instantiate(self.cfg.system.train.trainer, logger=TensorBoardLogger(cwd, '.', '.'),
                                                enable_model_summary=False,
                                               callbacks=[checkpoint, checkpoint2])
        trainer.fit(self.system, datamodule=datamodule)
```

### Error messages and logs

```
There is no error message, so it's very confusing.
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 2.2.3):
#- Lightning App Version (e.g., 0.5):
#- PyTorch Version (e.g., 2.1):
#- Python version (e.g., 3.8):
#- OS (e.g., Linux): Linux
#- CUDA/cuDNN version: CUDA
#- GPU models and configuration:
#- How you installed Lightning(`pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @awaelchli

**Comment by awaelchli** [comment]
@1zeryu Would you be able to provide a runnable code example that reproduces this please? You can use this template as a starting point:

```py
import os

import torch
from lightning.pytorch import LightningModule, Trainer
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()

```
Without it, it will be basically impossible for us to help. Thank you.

**Comment by 1zeryu** [comment]
Apologies for the confusing description, but the code is too long and complex,

**Comment by 1zeryu** [comment]
Here is my config of trainer:

**Comment by 1zeryu** [comment]
[code.zip](https://github.com/Lightning-AI/pytorch-lightning/files/14060009/code.zip)
Sorry for the confusing description, the code running by "python train_p_diff.py"

**Comment by 1zeryu** [comment]
Amazing I solve the bug by 
replacing:
loss = self.forward(batch, **kwargs)
self.optimizer.zero_grad()
self.manual_backward(loss)
self.optimizer.step()

with:
optimizer = self.optimizers()
loss = self.forward(batch, **kwargs)
optimizer.zero_grad()
self.manual_backward(loss)
optimizer.step()
